### PR TITLE
Adds some utility function usable to merge two databases together

### PR DESCRIPTION
**Pull request by Leusmann**

### CODE REVIEW

**Comment by tomsch420** [comment]
Write a better name and description

File: src/pycram/orm/utils.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import traceback
+
+import sqlalchemy
+import pycram.orm.base
+from pycram.designators.action_designator import *
+from pycram.designators.object_designator import *
+
+
+def logDB(in_session, filename, b_write_to_console=False):
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:9]: no abbreviations in variables, functions, etc..
Also no camel case for function names and variables.

Comment (tomsch420) [reply src/pycram/orm/utils.py:9]: Also type hint the parameters and return values

File: src/pycram/orm/utils.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    """
+        Utility function which stores a database in a local file.
+
+        :param in_session: (sqlalchemy.orm.session) Database Session which should be logged
+        :param filename: (String) Filename of the logfile
+        :param b_write_to_console: (bool) enables writing to the console. Default false
+        """
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:16]: wrong indent

File: src/pycram/orm/utils.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    f = open(filename, "w")
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:17]: with open(f, "w") as f:
instead of manual opening and closing

File: src/pycram/orm/utils.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    for table in Base.__subclasses__():
+        for column_object in in_session.query(table).all():
+            if b_write_to_console:
+                print("I am writing: {}".format(str(column_object.__dict__)))
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:21]: rospy.log

File: src/pycram/orm/utils.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            f.write(str(column_object.__dict__))
+            f.write("\n")
+    f.close()
+
+
+def printDB(in_SessionMaker):
+    """
+        Prints all ORM Class data within the given Session.
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:29]: wrong indent

File: src/pycram/orm/utils.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        """
+    memory_session = in_SessionMaker()
+    for table in Base.__subclasses__():
+        smt = sqlalchemy.select('*').select_from(table)
+        result = memory_session.execute(smt).all()
+        print(result)
+
+
+def update_primary_key(source_engine, destination_engine):
+    '''
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:41]: double quotes instead of single quotes in comment strings.

File: src/pycram/orm/utils.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        :param in_session: (sqlalchemy.orm.session) Database Session which should be printed
+        """
+    memory_session = in_SessionMaker()
+    for table in Base.__subclasses__():
+        smt = sqlalchemy.select('*').select_from(table)
+        result = memory_session.execute(smt).all()
+        print(result)
+
+
+def update_primary_key(source_engine, destination_engine):
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:40]: type hints

File: src/pycram/orm/utils.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    '''
+    Primary keys of destination will be updated so that there are no collisions when merging.
+    Care it is expected that the database takes care of cascading the change through its foreign keys
+    Will create own session
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:44]: Write a _coherent and consistent_ text in imperative form.

File: src/pycram/orm/utils.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    for table in Base.__subclasses__():  # iterate over all tables
+        highest_free_keyValue = 0
+        primaryKeys[table] = {}
+        list_of_primary_keys_of_this_table = table.__table__.primary_key.columns.values()
+        for key in list_of_primary_keys_of_this_table:
+            # make it smart but maybe
+            all_memory_key_values = memory_session.query(key).all()
+            primaryKeys[table][key.name] = real_session.query(key).all()
+            if all_memory_key_values:
+                highest_free_keyValue = max(all_memory_key_values)[0] + 1  # ToDo: Make it even more generic? # We will add 1 to the id then let the DB to the rest, afterwards we will compy all info over
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:61]: 80 digits limit for lines

File: src/pycram/orm/utils.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    Will create own session
+
+    :param source_engine: (sqlalchemy.orm.engine) Engine of the source data_base
+    :param destination_engine: (sqlalchemy.orm.engine) Engine of the destination data_base
+    '''
+    real_session = sqlalchemy.orm.Session(bind=destination_engine)
+    memory_session = sqlalchemy.orm.Session(bind=source_engine)
+    primaryKeys = {}
+    for table in Base.__subclasses__():  # iterate over all tables
+        highest_free_keyValue = 0
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:53]: snake_case

File: src/pycram/orm/utils.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        primaryKeys[table] = {}
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:54]: snake_case

File: src/pycram/orm/utils.py (hunk 9 above)
Comment (tomsch420) [inline src/pycram/orm/utils.py:61]: snake_case

File: src/pycram/orm/utils.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        list_of_primary_keys_of_this_table = table.__table__.primary_key.columns.values()
+        for key in list_of_primary_keys_of_this_table:
+            # make it smart but maybe
+            all_memory_key_values = memory_session.query(key).all()
+            primaryKeys[table][key.name] = real_session.query(key).all()
+            if all_memory_key_values:
+                highest_free_keyValue = max(all_memory_key_values)[0] + 1  # ToDo: Make it even more generic? # We will add 1 to the id then let the DB to the rest, afterwards we will compy all info over
+            for column_object in real_session.query(table).all():  # iterate over all columns
+                if column_object.__dict__[key.name] in all_memory_key_values:
+                    print("Found primarykey collision in table {} value: {} max value in memory {}".format(table,
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:64]: rospy.log

File: src/pycram/orm/utils.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                                                                                                           highest_free_keyValue))
+                    column_object.__dict__[key.name] = highest_free_keyValue
+                    highest_free_keyValue += 1
+
+    real_session.commit()
+    real_session.close()
+
+
+def copy_database(source_session_maker, destination_session_maker):
+    '''
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:76]: double qoutes and type hints

File: src/pycram/orm/utils.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    Primary keys of destination will be updated so that there are no collisions when merging.
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:77]: writing style

File: src/pycram/orm/utils.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+            for row in result:
+                for key in row:
+                    if not sqlalchemy.inspect(row[key]).detached and not sqlalchemy.inspect(
+                            row[key]).transient and not sqlalchemy.inspect(row[key]).deleted:
+                        source_session.refresh(row[key])
+                        source_session.expunge(row[key])
+                        sqlalchemy.orm.make_transient(row[key])
+                        objects_to_add.append(row[key])
+                    else:
+                        print("WARNING: Ignored already detached ORM Object {} ".format(
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:99]: rospy.warn

File: src/pycram/orm/utils.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+                            row[key]))  # ToDo: Maybe give option to include theses items (could break something)
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:100]: fix todo

Comment (tomsch420) [reply src/pycram/orm/utils.py:100]: Use Notes syntax for docstring to address these issues

**Review by tomsch420 — CHANGES_REQUESTED** [review]

**Comment by tomsch420** [comment]
Also create test cases

**Comment by Leusmann** [comment]
Should be fixed with the newest commits. Please check again.

File: src/pycram/orm/utils.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    destination_session.close()
+
+
+def copy_database(source_session_maker: sqlalchemy.orm.sessionmaker,
+                  destination_session_maker: sqlalchemy.orm.sessionmaker):
+    """
+    Iterates through all ORM Objects within tht source database and merges them into the destination database. Careful
+    this function does not check if there are any primary key collisions or updates any data.
+
+    note:
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:87]: brah

File: src/pycram/orm/utils.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+        rospy.loginfo(result)
+
+
+def update_primary_key(source_engine: sqlalchemy.orm.engine, destination_engine: sqlalchemy.orm.engine):
+    """
+    Updates all the primary keys of the database associated with the destination engine, so that there will be no
+    problems when merging it into the source database. In order to achieve this the highest id value of the source
+    engine is searched and the primary keys of the destination database will get all the values following that.
+    Cascading triggers in the database will take care of the rest. Careful as today (20.09.2023) this will not work in
+    memory databases as there are no triggers
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:46]: as of version x.x

File: src/pycram/orm/utils.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+from pycram.designators.action_designator import *
+from pycram.designators.object_designator import *
+
+
+def write_database_to_file(in_session: sqlalchemy.orm.session, filename: str, b_write_to_console: bool = False):
+    """
+    Writes all ORM Objects stored within the given session into a local file.
+
+    :param in_session: (sqlalchemy.orm.session) Database Session which should be logged
+    :param filename: (String) Filename of the logfile
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:15]: duplicated information

File: src/pycram/orm/utils.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
+import traceback
+
+import rospy
+import sqlalchemy
+import pycram.orm.base
+from pycram.designators.action_designator import *
+from pycram.designators.object_designator import *
+
+
+def write_database_to_file(in_session: sqlalchemy.orm.session, filename: str, b_write_to_console: bool = False):
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:10]: This hints a module and not the Session class

**Review by tomsch420 — CHANGES_REQUESTED** [review]

**Comment by Leusmann** [comment]
I would put the testcases in another merge request but it is up to you

**Comment by tomsch420** [comment]
No merge without testing

File: test/test_database_merger.py [hunk 21]
```diff
@@ -16,21 +18,21 @@
 from pycram.bullet_world import BulletWorld, Object
 from pycram.designators.object_designator import *
 import anytree
-
+from anytree import Node,RenderTree,LevelOrderIter
```
Comment (tomsch420) [inline test/test_database_merger.py:21]: ctrl+alt+shift+L

File: test/test_database_merger.py [hunk 22]
```diff
@@ -39,43 +41,56 @@ def pick_and_place_plan(self):
...
-#Note: Can't test full functionallity
+# Note: Can't test full functionallity
 
 class MergeDatabaseTest(unittest.TestCase):
     source_engine: sqlalchemy.engine.Engine
     destination_engine: sqlalchemy.engine.Engine
     source_session_maker: sqlalchemy.orm.sessionmaker
     destination_session_maker: sqlalchemy.orm.sessionmaker
 
+    @unittest.skipIf(not (pathlib.Path("test_database_merger.json").resolve().is_file()),
```
Comment (tomsch420) [inline test/test_database_merger.py:78]: make the configuration a class.

File: test/test_database_merger.py [hunk 23]
```diff
@@ -122,14 +137,17 @@ def TearDownClass(cls):
...
+        ordered_orm_classes = [node.name for node in LevelOrderIter(orm_tree)]
+        pycram.orm.utils.update_primary_key_constrains(self.destination_session_maker,ordered_orm_classes)
+        pycram.orm.utils.update_primary_key(self.source_session_maker,
+                                            self.destination_session_maker)  # self.source_engine,self.destination_engine)
+        # pycram.orm.utils.print_database(self.destination_session_maker)
+        # print("Source:" + 10 * '-')
+        # pycram.orm.utils.print_database(self.source_session_maker)
+        # pycram.orm.utils.copy_database(self.destination_session_maker,self.source_session_maker)
         pycram.orm.utils.copy_database(self.source_session_maker, self.destination_session_maker)
+        pycram.orm.utils.print_database(self.source_session_maker)
```
Comment (tomsch420) [inline test/test_database_merger.py:151]: assert what?

File: src/pycram/orm/utils.py [hunk 24]
```diff
@@ -37,44 +38,86 @@ def print_database(in_Sessionmaker: sqlalchemy.orm.sessionmaker):
...
+        tree=node
+    else:
+        node=Node(in_node,parent)
+    if len(in_node.__subclasses__()):
+        for subnode in in_node.__subclasses__():
+            get_tree(subnode,node,tree)
+    return tree
+
+
+def update_primary_key(source_session_maker: sqlalchemy.orm.sessionmaker,
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:67]: functions should do one thing. This function can be partitioned in a lot more functions

File: src/pycram/orm/utils.py [hunk 25]
```diff
@@ -84,40 +127,83 @@ def copy_database(source_session_maker: sqlalchemy.orm.sessionmaker,
...
+    endpoint to on update cascading. Careful currently only works on postgres databases.
+    :param session_maker:
+    :param orm_classes:
+    :return: empty
+    '''
+    with session_maker() as session:
+        for orm_class in orm_classes:
+            try:
+                foreign_key_statement = sqlalchemy.text(
+                    "SELECT con.oid, con.conname, con.contype, con.confupdtype, con.confdeltype, con.confmatchtype, pg_get_constraintdef(con.oid) FROM pg_catalog.pg_constraint con INNER JOIN pg_catalog.pg_class rel ON rel.oid = con.conrelid INNER JOIN pg_catalog.pg_namespace nsp ON nsp.oid = connamespace WHERE rel.relname = '{}';".format(
```
Comment (tomsch420) [inline src/pycram/orm/utils.py:187]: replace with sqlalchemy queries.

**Review by tomsch420 — CHANGES_REQUESTED** [review]

**Comment by Leusmann** [comment]
Did some update here. Please check again.

File: examples/migrate_neems.ipynb [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "19ffb917198a0ed9",
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "i=0\n",
```
Comment (tomsch420) [inline examples/migrate_neems.ipynb:49]: that cell seems unnescessary

File: examples/migrate_neems.ipynb [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython2",
+   "version": "2.7.6"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 5
+}
```
Comment (tomsch420) [inline examples/migrate_neems.ipynb:174]: show the cool output of the merged database

**Review by tomsch420 — APPROVED** [review]

**Pull request merged**